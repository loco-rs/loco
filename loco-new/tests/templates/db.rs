use loco::{settings, wizard::DBOption};
use rstest::rstest;

use super::*;
use crate::assertion;

pub fn run_generator(db: DBOption) -> TestGenerator {
    let settings = settings::Settings {
        package_name: "loco-app-test".to_string(),
        module_name: "loco_app_test".to_string(),
        db: db.into(),
        ..Default::default()
    };

    TestGenerator::generate(settings)
}

#[rstest]
fn test_config_file_no_db(
    #[values("config/development.yaml", "config/test.yaml")] config_file: &str,
) {
    let generator = run_generator(DBOption::None);
    let content = assertion::yaml::load(generator.path(config_file));
    assertion::yaml::assert_path_is_empty(&content, &["database"]);
}

#[rstest]
fn test_config_with_sqlite(
    #[values(DBOption::Sqlite, DBOption::Postgres)] db: DBOption,
    #[values("config/development.yaml", "config/test.yaml")] config_file: &str,
) {
    let generator = run_generator(db.clone());
    let content = assertion::yaml::load(generator.path(config_file));

    insta::assert_snapshot!(
        format!(
            "{}_config_database_{:?}",
            config_file.replace(['/', '.'], "_"),
            db
        ),
        format!(
            "{:#?}",
            assertion::yaml::get_value_at_path(&content, &["database"]).unwrap()
        )
    );
}

#[rstest]
fn test_cargo_toml(#[values(DBOption::None, DBOption::Sqlite, DBOption::Postgres)] db: DBOption) {
    let generator = run_generator(db.clone());
    let content = assertion::toml::load(generator.path("Cargo.toml"));

    insta::assert_snapshot!(
        format!("cargo_dependencies_{:?}", db),
        content.get("dependencies").unwrap()
    );
}

#[rstest]
fn test_app_rs(#[values(DBOption::None, DBOption::Sqlite, DBOption::Postgres)] db: DBOption) {
    let generator = run_generator(db.clone());
    insta::assert_snapshot!(
        format!("src_app_rs_{:?}", db),
        std::fs::read_to_string(generator.path("src/app.rs")).expect("could not open file")
    );
}

#[rstest]
fn test_src_lib_rs(#[values(DBOption::None, DBOption::Sqlite, DBOption::Postgres)] db: DBOption) {
    let generator = run_generator(db.clone());

    let content =
        std::fs::read_to_string(generator.path("src/lib.rs")).expect("could not open file");

    if db.enable() {
        assertion::string::assert_line_regex(&content, "(?m)^pub mod models;$");
    } else {
        assertion::string::assert_str_not_exists(&content, "pub mod models;");
    }
}

#[rstest]
fn test_src_bin_main_rs(
    #[values(DBOption::None, DBOption::Sqlite, DBOption::Postgres)] db: DBOption,
) {
    let generator = run_generator(db.clone());

    let content =
        std::fs::read_to_string(generator.path("src/bin/main.rs")).expect("could not open file");

    if db.enable() {
        assertion::string::assert_line_regex(&content, "(?m)^use migration::Migrator;$");
        assertion::string::assert_line_regex(
            &content,
            r"(?m)^    cli::main::<App, Migrator>\(\).await$",
        );
    } else {
        assertion::string::assert_str_not_exists(&content, "(?m)^use migration::Migrator;$");
        assertion::string::assert_line_regex(&content, r"(?m)^    cli::main::<App>\(\).await");
    }
}

#[rstest]
#[cfg(windows)]
fn test_src_bin_tool_rs(
    #[values(DBOption::None, DBOption::Sqlite, DBOption::Postgres)] db: DBOption,
) {
    let generator = run_generator(db.clone());

    let content =
        std::fs::read_to_string(generator.path("src/bin/tool.rs")).expect("could not open file");

    if db.enable() {
        assertion::string::assert_line_regex(&content, "(?m)^use migration::Migrator;$");
        assertion::string::assert_line_regex(
            &content,
            r"(?m)^    cli::main::<App, Migrator>\(\).await$",
        );
    } else {
        assertion::string::assert_str_not_exists(&content, "(?m)^use migration::Migrator;$");
        assertion::string::assert_line_regex(&content, r"(?m)^    cli::main::<App>\(\).await");
    }
}

#[rstest]
fn test_tasks_mod_rs(#[values(DBOption::None, DBOption::Sqlite, DBOption::Postgres)] db: DBOption) {
    let generator = run_generator(db.clone());

    let content =
        std::fs::read_to_string(generator.path("src/tasks/mod.rs")).expect("could not open file");

    if db.enable() {
        assertion::string::assert_line_regex(&content, "(?m)^pub mod seed;$");
    } else {
        assertion::string::assert_str_not_exists(&content, "pub mod seed");
    }
}

#[rstest]
fn test_tests_mod_rs(#[values(DBOption::None, DBOption::Sqlite, DBOption::Postgres)] db: DBOption) {
    let generator = run_generator(db.clone());

    let content =
        std::fs::read_to_string(generator.path("tests/mod.rs")).expect("could not open file");

    if db.enable() {
        assertion::string::assert_line_regex(&content, "(?m)^mod models;$");
    } else {
        assertion::string::assert_str_not_exists(&content, "mod models;");
    }
}

#[rstest]
fn test_tests_tasks_mod_rs(
    #[values(DBOption::None, DBOption::Sqlite, DBOption::Postgres)] db: DBOption,
) {
    let generator = run_generator(db.clone());

    let content =
        std::fs::read_to_string(generator.path("tests/tasks/mod.rs")).expect("could not open file");

    if db.enable() {
        assertion::string::assert_line_regex(&content, "(?m)^pub mod seed;$");
    } else {
        assertion::string::assert_str_not_exists(&content, "pub mod seed");
    }
}
