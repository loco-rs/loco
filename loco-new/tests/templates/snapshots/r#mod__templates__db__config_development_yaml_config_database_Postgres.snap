---
source: tests/templates/db.rs
expression: "format!(\"{:#?}\",\nassertion::yaml::get_value_at_path(&content, &[\"database\"]).unwrap())"
snapshot_kind: text
---
Mapping {
    "uri": Mapping {
        Mapping {
            "get_env(name=\"DATABASE_URL\"": Null,
            "default=\"postgres://loco:loco@localhost:5432/loco-app-test_development\")": Null,
        }: Null,
    },
    "enable_logging": Bool(false),
    "connect_timeout": Mapping {
        Mapping {
            "get_env(name=\"DB_CONNECT_TIMEOUT\"": Null,
            "default=\"500\")": Null,
        }: Null,
    },
    "idle_timeout": Mapping {
        Mapping {
            "get_env(name=\"DB_IDLE_TIMEOUT\"": Null,
            "default=\"500\")": Null,
        }: Null,
    },
    "min_connections": Mapping {
        Mapping {
            "get_env(name=\"DB_MIN_CONNECTIONS\"": Null,
            "default=\"1\")": Null,
        }: Null,
    },
    "max_connections": Mapping {
        Mapping {
            "get_env(name=\"DB_MAX_CONNECTIONS\"": Null,
            "default=\"1\")": Null,
        }: Null,
    },
    "auto_migrate": Bool(true),
    "dangerously_truncate": Bool(false),
    "dangerously_recreate": Bool(false),
}
